File: uapc/archive/2022/problems/ordinaryordinals/submissions/accepted/ordinaryordinals-jc.py
#!/usr/bin/python3

# using built-in pow(a, b, m) would also work.
def binpow(a, b, m):
    res = 1
    while b > 0:
        if b & 1: res = res*a%m
        a = a*a%m
        b >>= 1
    return res

n, m = map(int, input().split())
print((binpow(2, n+1, m) + binpow(2, n-1, m) - 1)%m)

File: uapc/archive/2022/problems/ordinaryordinals/submissions/accepted/ordinaryordinals-jc.cpp
#include <iostream>

using namespace std;
typedef unsigned long long ll;

ll binpow(ll a, ll b, ll m) {
    ll res = 1;
    while (b > 0) {
        if (b & 1) res = res*a%m;
        a = a*a%m;
        b >>= 1;
    }
    return res;
}

int main() {
    ll n, m;
    cin >> n >> m;
    cout << (n ? (binpow(2, n+1, m)+binpow(2, n-1, m)-1) : 2) % m << endl;
    return 0;
}
